internal/querycoordv2/balance/score_based_balancer.go

```diff
@@ -232,91 +248,65 @@ func (b *ScoreBasedBalancer) genStoppingSegmentPlan(replica *meta.Replica, onlin
-		fromSegments = lo.Filter(fromSegments, func(segment *meta.Segment, _ int) bool {
-			return segment.GetLevel() != datapb.SegmentLevel_L0
-		})
-		sort.Slice(fromSegments, func(i, j int) bool {
-			return fromSegments[i].GetNumOfRows() < fromSegments[j].GetNumOfRows()
-		})
-		var targetSegmentToMove *meta.Segment
-		for _, segment := range fromSegments {
-			targetSegmentToMove = segment
-			if havingMovedSegments.Contain(targetSegmentToMove.GetID()) {
-				targetSegmentToMove = nil
-				continue
-			}
-			break
-		}
-		if targetSegmentToMove == nil {
-			// the node with the highest score doesn't have any segments suitable for balancing, stop balancing this round
-			break
+		rowCount := b.calculateScore(replica.CollectionID, node)
+		totalScore += rowCount
+		nodeScore[node] = rowCount
+	}
+
+	if totalScore == 0 {
+		return nil
+	}
+
+	// find the segment from the node which has more score than the average
+	segmentsToMove := make([]*meta.Segment, 0)
+	average := totalScore / len(onlineNodes)
+	for node, segments := range segmentDist {
+		leftScore := nodeScore[node]
+		if leftScore <= average {
+			continue
 		}
 
-		nextFromPriority := fromPriority - int(targetSegmentToMove.GetNumOfRows()) - int(float64(targetSegmentToMove.GetNumOfRows())*
-			params.Params.QueryCoordCfg.GlobalRowCountFactor.GetAsFloat())
-		nextToPriority := toPriority + int(targetSegmentToMove.GetNumOfRows()) + int(float64(targetSegmentToMove.GetNumOfRows())*
-			params.Params.QueryCoordCfg.GlobalRowCountFactor.GetAsFloat())
-
-		// still unbalanced after this balance plan is executed
-		if nextToPriority <= nextFromPriority {
-			plan := SegmentAssignPlan{
-				ReplicaID: replica.GetID(),
-				From:      fromNode.nodeID,
-				To:        toNode.nodeID,
-				Segment:   targetSegmentToMove,
-			}
-			segmentPlans = append(segmentPlans, plan)
-		} else {
-			// if unbalance reverted after balance action, we will consider the benefit
-			// only trigger following balance when the generated reverted balance
-			// is far smaller than the original unbalance
-			nextUnbalance := nextToPriority - nextFromPriority
-			if float64(nextUnbalance)*params.Params.QueryCoordCfg.ReverseUnbalanceTolerationFactor.GetAsFloat() < unbalance {
-				plan := SegmentAssignPlan{
-					ReplicaID: replica.GetID(),
-					From:      fromNode.nodeID,
-					To:        toNode.nodeID,
-					Segment:   targetSegmentToMove,
-				}
-				segmentPlans = append(segmentPlans, plan)
-			} else {
-				// if the tiniest segment movement between the highest scored node and lowest scored node will
-				// not provide sufficient balance benefit, we will seize balancing in this round
+		sort.Slice(segments, func(i, j int) bool {
+			return segments[i].GetNumOfRows() < segments[j].GetNumOfRows()
+		})
+		for _, s := range segments {
+			segmentsToMove = append(segmentsToMove, s)
+			leftScore -= b.calculateSegmentScore(s)
+			if leftScore <= average {
 				break
 			}
```

Comment:
these code may case node score below avg score and cause segment bouncing between nodes

Reply:
we calculate the benefit during the `assignSegment`, if  reassign won't get enough benefits, we will skip this reassignment